Restructure Contribute and Integrate & Extend sections.

BenLloydPearson opened the pull request:
#### Summary

This is a little clunky, but merging these sections required a lot of little changes in a ton of different files.

Here's a summary of everything these commits do:
* Created 2 new pages for contribute/getting-started and /integrate/getting-started
* Refactored all instances of non-image URLs containing` /extend/ to /integrate/
* Moved all child content from Extend section to Integrate, created aliases for every document that was moved to handle redirects.
* Replaced Extend on the homepage with Development Guides
* Removed Extend from top nav
* Reorganized items in the left nav bars for contribute and integrate sections.

I've tested everything locally, and it all seems to be working fine. But, I'd appreciate a 2nd set of eyes to make sure I haven't missed anything important.



#### Ticket Link
n/a

mickmister commented on site/content/integrate/plugins/server/reference.md:
```diff
@@ -4,6 +4,7 @@ heading: "Mattermost Server Reference"
 description: "The plugin package is used by Mattermost server plugins written in Go and enables you to manage and interact with the plugin environment."
 date: 2018-07-10T00:00:00-05:00
 weight: 10
+aliases: [/extend/plugins/server/reference/]
```
Is this used to support existing links to the `extend` subpath?

BenLloydPearson replied:
@mickmister If I understand your answer correctly, yes. Aliases in Hugo create automatic redirects for the file path to send traffic to this page. I tested these locally and they perform as expected.

mickmister commented on site/content/contribute/getting-started/_index.md:
```diff
@@ -1,24 +1,32 @@
+## Technical Overview
 
-{{< bignumber number="1" title="Sign up to our Mattermost site" content="Go to [community.mattermost.com](https://community.mattermost.com/signup_user_complete) and join the [Developers](https://community.mattermost.com/core/channels/developers) and [Contributors](https://community.mattermost.com/core/channels/tickets) channels to ask questions." >}}
+The Mattermost core repositories include:
+* [Server](https://developers.mattermost.com/contribute/server/) - Highly-scalable Mattermost installation written in Go
+* [Web App](https://developers.mattermost.com/contribute/webapp/) - JavaScript client app built on React and Redux
+* [Mobile Apps](https://developers.mattermost.com/contribute/mobile/) - JavaScript client apps for Android and iOS built on React Native
+* [Desktop App](https://developers.mattermost.com/contribute/desktop/) - An Electron wrapper around the web app project that runs on Windows, Linux, and macOS
```
This section is titled as "The Mattermost core repositories", but can we have a mention of contributing to plugin repositories on this landing page as well?

BenLloydPearson replied:
@mickmister I added a link to /contribute/plugins to this list. That page is light on content right now, but I can improve it in my next PR. I would like to highlight the plugins we maintain a lot more.

mickmister commented on site/content/contribute/server/plugins.md:
```diff
@@ -12,17 +12,17 @@ This document covers the plugin infrastructure and how to contribute to it.
 
 ## Building Plugins
 
```
Maybe this page should be in the `contribute/plugins` instead of `contribute/server`?

BenLloydPearson replied:
This page definitely seems out of place, but I'd like to focus this PR on solving the issue of merging Integrate and Extend. I would like to follow up with a second pass to review all of the pages in the left sidebar to see if they need further restructuring, deprecation, or updates, and this page will almost certainly be included in that.

mickmister reviewed (approved):
Awesome work @BenLloydPearson! It always bothers me that there are separate `integrate` and `extend` sections, when they should really be combined into one section as you've done here.

I have some questions below, and a non-blocking request to include a mention of contributing to plugin repositories on the "Getting Started" developer contribution landing page.

cwarnermm commented:
@BenLloydPearson - This is a really great PR. Thank you!! 

There's one thing I've noticed throughout the PR that we should update before merging. Reference links in the following format should be changed to Hugo-friendly ref links so that they can check and resolve automatically:

``The plugin [developer setup](/integrate/plugins/developer-setup/) and [developer workflow](/integrate/plugins/developer-workflow/) pages are useful to learn about the plugin development environment. You can find more information about plugins in general [here](/integrate/plugins/).``

See [this page](https://github.com/mattermost/mattermost-developer-documentation/edit/master/site/content/integrate/admin-guide/admin-interactive-dialogs/_index.md) as an example of how to format the ref links. Links formatted this way don't require us to change the link later if things move around. The ref is then looking for the MD file, rather than the path. 

CC @justinegeffen @hanzei - Please correct me if I've mis-communicated our preferred link approach in the developer documentation set.

cwarnermm reviewed (approved):
This PR has great improvements for the developer product documentation. Thank you!

BenLloydPearson commented:
@cwarnermm I updated the new content that I generated with the appropriate markdown file references, but some of them are links to _index.md files inside directories and I'm not sure how to handle those. I also spotted a few links I needed to update to work with the new other-integrations sub-directory that I created. 

However, many of the pages that I modified to update the URLs aren't currently following your recommendations. I agree that we should be consistent, but is that something we should create a separate issue to resolve? There are potentially hundreds of links that we'd need to review.

justinegeffen commented:
@BenLloydPearson, there are definitely a ton of inconsistencies in the docs which is something we should address long-term. What I've been trying to do is, going forward, use the Hugo ref as per @hanzei's recommendation. We will eventually get all the internal links changed to refs but that's a whole other set of work.

You mention that some are links to _index.md files inside directories - I don't actually know what the ref would be (good catch!). @hanzei, can you help out here? 

Otherwise, this is an awesome PR and much-appreciated. :)

cwarnermm commented:
@BenLloydPearson - Agreed completely that there are likely hundreds of problematic links worth updating. That's definitely not on your a requirement of this PR. 

However, if you're open to it, it would be great if the links within the scope of these PRs were updated based on @hanzei's recommendations.

BenLloydPearson commented:
@cwarnermm I've updated all of the links that are affected by this PR to the appropriate file references. I also realized that I forgot to create content for the new Other Integrations page, and I also corrected this.

I think this should address all of the concerns raised in this PR.


P.s. for _index.md files, the references look like this: ({{< ref "/integrate/" >}}) to get the _index.md inside the integrate directory.

justinegeffen commented:
@BenLloydPearson, there are a bunch of conflicts that have popped up due to the Docathon work. They should mainly be URL updates but they seem like they may be complex as GitHub requires them to be resolved in the command line. Let me know if you need any help with this.

justinegeffen commented:
I think I can see what happened - all the docathon changes were full URLs but these docs have all had the (correct) references added in the interim. Not sure how to resolve that easily. :(

BenLloydPearson commented:
@justinegeffen This should be ready to be merged now. One of the link changes is no longer relevant because it was removed in my new version. The other has been updated.
